### DISCUSSION TITLE
Outlet off, delay, back on

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#17508] [General]
Hi,

Happy holidays to everyone :). 

While waiting for our guests to get back, I started playing with my tasmota devices and realized I sure could use a way to power cycle certain devices. 

I found this;
Backlog Status 1; Power2 on; Delay 20; Power2 off; Status 4

But I'm not sure if this is a reliable way to do it or if this is even what I'm looking for. Obviously, this one shows on then off but I need off then on.

**Comment by sfromis** [comment]
If you set `PowerOnState 5`, and `PulseTime 20`, Tasmota can handle your power cycle timing automatically and reliably, meaning that Tasmota will turn power on again two seconds after you power off by whatever means. The downside may be that you cannot turn off power for good.

Using `Delay` in that way is not a reliable way of ensuring timing, as the interval can be be cut short. For reliable timing, you can use rules with RuleTimer, if you do not want `PowerOnState 5`

**Comment by dac2020** [comment]
Hi,

That might be a problem since I would need to have full control of on or off but definitely could use a way to power cycle devices now and then from remote.

However, what if I sent it a series of commands? 
If I wanted to power cycle something, I could send the above commands, wait 30 seconds then reset those back to default?

**Reply by sfromis** [reply]
`Delay` is not the way to go if you want reliable timing (without the risk of being cut short. Obviously, if you are willing to do a command, wait, and then another command, you can just send power off/on commands. This would not work if the thing to be cycled controls network traffic. Still, the `Delay` would work a lot of the time, just not if you want it to be reliable.

Here's a way to use my recommendation of RuleTimer for a rule set without the risk of being cut short:
```
Rule1 on Event#PowerCycle do Backlog Power1 off; RuleTimer1 2 endon on Rules#Timer=1 do Power1 on endon
Rule1 1
```

And then `Event PowerCycle` as a command.

**Comment by dac2020** [comment]
The problem is that I can only send one command. 

For example, if I send a command to a remote to reboot the router for example, the device would need to have the commands before the Internet connection was lost.

So that single command has to contain not only a shut off outlet command but 10 seconds later, a turn on one as well otherwise, remote access is lost.

Hope I'm understanding what you are suggesting. 
Do you mean create a rule as above and then send some sort of command to run that rule?

**Reply by sfromis** [reply]
That was **exactly** why I posted the rule set allowing for that. When you once has created that rule set, the Event command triggers the action, without any other preparation each time.

Of course, the power on cannot happen if power is cut from the Tasmota device itself during the 2 seconds (you could change to 10 if you want that instead).

**Comment by dac2020** [comment]
I understand. Just not very familiar with rule sets as the only thing I've ever done with those is remove some.
Time to read up and take your suggestion.

Thank you.

**Comment by dac2020** [comment]
I've been playing with this and it seems to work fine.

For example, on a two outlet device.

Rule1 on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 4 endon on Rules#Timer=1 do Power1 on endon
Rule1 1
Rule2 on Event#PowerCycle2 do Backlog Power2 off; RuleTimer1 4 endon on Rules#Timer=1 do Power2 on endon
Rule2 1

Event PowerCycle1 - cycles outlet 1
Event PowerCycle2 - cycles outlet 2

Is this the best way of doing it? It means creating 8 rules for an 8 outlet device for example. 
Any memory constrains for such a number of rules?

The only odd thing so far is that sometimes, sending the command, nothing happens at all or for quite a while then it suddenly runs.

**Reply by sfromis** [reply]
One rule set for up to 8 power channels.
`on Event#PowerCycle do Backlog Power%value% off; RuleTimer%value% 4 endon on Rules#Timer do Power%value% on endon`

And then:
`Event PowerCycle=2`

A limitation is that you cannot easily combine it with other usage of the 8 RuleTimer available, as each is used for the power channel with the same number.

Still, 8 times the two rules would be no memory problem, as you have 3 x 512 bytes, which due to compression can hold more than 3KB of rule text.

**Reply by dac2020** [reply]
What is happening after using multiple rules is that I'll send an event and it works one time, not the next or not consistently. 
One time, it might never run the cycle event at all and another time, after a very long delay, it runs but it then changes the status of other outlets.

**Reply by dac2020** [reply]
I removed those rules.
I added;

Console
 on Event#PowerCycle do Backlog Power%value% off; RuleTimer%value% 4 endon on Rules#Timer do Power%value% on endon
 stat/B4E62D773B84/RESULT = {"Command":"Unknown"}

I think I'm taking your command too literally :). Have to go check the commands page.

**Reply by sfromis** [reply]
You should of course have the usual stuff with `Rule1` and so on to update the rule set. I was only mentioning the rules, not the commands to update and enable the rule set.

**Reply by sfromis** [reply]
Your original set of rules had one bad error, you were using RuleTimer1 for both, meaning that if you power cycled both within 4 seconds, then 2nd one would not get a chance to turn on again.

**Reply by dac2020** [reply]
Ah, that explains the outlets seemingly going crazy. 
I'll re-add the rules and review here as I do.

**Reply by sfromis** [reply]
Just do not worry about that, my suggestion handles up to 8 power channels with one single set of two rules.

**Reply by dac2020** [reply]
I re-ran the rules using
Rule1 on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 5 endon on Rules#Timer=1 do Power1 on endon
Rule1 1
Rule2 on Event#PowerCycle2 do Backlog Power2 off; RuleTimer2 5 endon on Rules#Timer=2 do Power2 on endon
Rule2 1

I did this for all the outlets I want to cycle and each works perfectly now.

**Reply by sfromis** [reply]
Sure, when fixing the error, it works better. But you asked for a better way of doing it, and I gave you one, where you do not need separate rules for each power channel.

**Reply by dac2020** [reply]
Yes, I have that too now. 

Thank you very much for your help.

**Reply by dac2020** [reply]
Can commands be stacked?

For example

Rule1 on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 5 endon on Rules#Timer=1 do Power1 on endon
Rule1 1

Could this all be one line?

**Reply by sfromis** [reply]
Stacking can be done using Backlog, but you cannot have Backlog within that. If you want to experiment, the command `JSON` exists, taking commands via a JSON data structure, like:
`json {"backlog":"var1 0;var2 0","backlog":"add1 1;add2 1"}`
One command, an still using Backlog twice. This feature is not designed for such tricks, and there may be some gotcha somewhere.

In some cases, you can also avoid Backlog by having the same trigger used on a separate rule for each command.

**Reply by dac2020** [reply]
I found something odd. 
On a power bar, I want to configure only the first 4 outlets to power cycle.
When doing so, I am only able to enter three of these rules otherwise, I get an error.

```
14:44:49.550 CMD: Rule1 on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 5 endon on Rules#Timer=1 do Power1 on endon
14:44:49.557 RUL: Stored uncompressed, would compress from 98 to 72 (-27%)
14:44:49.605 MQT: stat/B4E62D773B84/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 5 endon on Rules#Timer=1 do Power1 on endon"}}

14:44:52.802 CMD: Rule2 on Event#PowerCycle2 do Backlog Power2 off; RuleTimer2 5 endon on Rules#Timer=2 do Power2 on endon
14:44:52.809 RUL: Stored uncompressed, would compress from 98 to 72 (-27%)
14:44:52.856 MQT: stat/B4E62D773B84/RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle2 do Backlog Power2 off; RuleTimer2 5 endon on Rules#Timer=2 do Power2 on endon"}}

14:44:58.001 CMD: Rule3 on Event#PowerCycle3 do Backlog Power3 off; RuleTimer3 5 endon on Rules#Timer=3 do Power3 on endon
14:44:58.007 RUL: Stored uncompressed, would compress from 98 to 72 (-27%)
14:44:58.049 MQT: stat/B4E62D773B84/RESULT = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle3 do Backlog Power3 off; RuleTimer3 5 endon on Rules#Timer=3 do Power3 on endon"}}

14:44:44.630 CMD: Rule0
14:44:44.751 MQT: stat/B4E62D773B84/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle1 do Backlog Power1 off; RuleTimer1 5 endon on Rules#Timer=1 do Power1 on endon"}}
14:44:44.813 MQT: stat/B4E62D773B84/RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle2 do Backlog Power2 off; RuleTimer2 5 endon on Rules#Timer=2 do Power2 on endon"}}
14:44:44.889 MQT: stat/B4E62D773B84/RESULT = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":98,"Free":413,"Rules":"on Event#PowerCycle3 do Backlog Power3 off; RuleTimer3 5 endon on Rules#Timer=3 do Power3 on endon"}}

14:45:01.818 CMD: Rule4 on Event#PowerCycle4 do Backlog Power4 off; RuleTimer4 5 endon on Rules#Timer=4 do Power4 on endon
14:45:01.908 MQT: stat/B4E62D773B84/RESULT = {"Command":"Error"}
```
I can control the other outlets manually using Powerx 1/0 so I know they are working.

**Reply by sfromis** [reply]
You only have 3 rule *sets*, but each set can have a **long** list of rules. You could have it all, and much more in rule set number 1, without using rule sets 2 and 3.

**Reply by dac2020** [reply]
Ok, then I'm back to your single command.

Rule1 on Event#PowerCycle do Backlog Power%value% off; RuleTimer%value% 5 endon on Rules#Timer do Power%value% on endon
Rule1 1
Event PowerCycle=1

Works as expected, thank you.

Question. Can this command be used on devices with different numbers of outlets, say 1 to 8 for example? It seems so but just want to confirm.

**Reply by sfromis** [reply]
That was a lot of editing for a 2nd version of the post :grin: 

Yes, it works up to 8, as the number of available RuleTimer is limited to 8. It would get a lot more complex to manage it reliably for more than 8 channels, if they are to operate at the same time without getting mixed up. What makes my suggestion simple is using the same number for the power channel, and for the RuleTimer.

Anyway, it can also work just fine with rules including specific numbers, if you prefer. You could simply string a long list of rules into one single rule set number 1, easily having 10 or much more rules in each of the 3 available rule sets.

You should be aware of what rules vs rule sets are. Each rule is from `on` to `endon`, only. You can then have a long list of rules in the same rule set, simply by having the rules one after the other. And then you have the command `Rule` for updating each of the 3 rule sets. Most of the time, each rule is independent of other rules, there is no coupling by being within the same rule set, except for setting special options like one-shot, or using `break`.

**Reply by dac2020** [reply]
Maybe more than two versions. I tried, made a mistake, was posting, tried again, fixed it but found another error, etc :).
I'll make note of all this.

BTW, someone shared a while back how to clear all rules, something like rule0 '' maybe?
I created at least three rules but now only need the one you've suggested.

Not understanding the doc and how you clear rules.
Rule<x>
" = clear Rule<x>

For example;

Rule2 ''
Does not clear the rule.

**Reply by sfromis** [reply]
> `something like rule0 ''`
No, `rule0 "`

Used monospace font to hilight the important difference

And it does not just clear one rule, but a whole rule set. And with `Rule0`, all 3 rule sets.

**Reply by dac2020** [reply]
Got it, thank you.
